### DISCUSSION TITLE
굳이 클래스로 만들 필요가 있을까요

### DISCUSSION THREAD

**Discussion by moonheekim0118** [woowacourse-study/2022-code-review-study#36] [Lotto] [answered]
매우매우 심오한.. 질문인데요, 크루들 보면 다 클래스를 사용해서 프로그래밍하는걸 볼 수 있어요. 저도 마찬가지구요.
그런데..살펴보면 그 클래스로 인스턴스는 대부분 딱 한번 생성되거든요. 여러번 반복되어 재사용되면 몰라도, 딱 한 번사용되는데 왜 우리는 클래스를 사용하고 있는걸까요 ? 너무 심오한가요..? 그냥 갑자기 든 의문입니다. 다들 딱 한번 인스턴스 생성되고 말 클래스를 사용하는 뚜렷한 이유가 있을까요?  여러분의 의견이 궁금해요~

**Comment by lokba** [comment]
오호..

아래에 마르코가 남긴 `적합한 설계가 무엇인가요?`에서 제가 머릿속으로 고민했던 부분이랑 똑같아요!!

저 또한, 고민을 해보니 호프랑 동일한 생각을 하였는데요.

제 기억으로는,
프리코스 숫자 야구게임 피드백에서 `class를 사용해라`가 있었어요.
왜 코치들은, 프리코스때부터 class를 언급했을까요?


- 이유 1
js는 class를 지원하고 class와 관련된 많은 문법들을 최근에 지원하고 있어요.
이것은,  `객체지향적 프로그래밍이 선호된다`라고 해석이 될 수 있을 것 같아요.


- 이유 2
또 다른이유로는,
level2 부터, 리액트를 사용하자나요.
리액트또한 컴포넌트를 class기반, 함수형 기반으로 나뉘게 되는데,
함수형 기반이 class 기반보다 코드 짜기도 편하고 상태관리도 수월하고 여러 이점이 많아서 함수형 기반을 많이 선호하고,
그리해서 level2도 높은 확률로 함수형 기반으로 진행되지 않을까 싶어요. 
만약, 그렇다면 어차피 `함수형으로 짜니까 지금은 class에 더 몰입해` 일수도 있을것 같아요ㅋㅋ(사실 억지로 이유를 만든거여서....)

**Comment by wonsss** [comment, marked as answer]
인스턴스가 1개만 생성되고, 프로그램 종료시까지 해당 인스턴스만 사용한다면
class를 사용하지 않아도 동일 기능을 구현 가능할 것 같아요.

아래처럼 class의 기능을 전역변수와 함수로 대체할 수 있지 않을까요?
- class -> js 파일 단위
- class 멤버변수 -> 해당 js 파일의 전역변수
- class 메서드 -> 해당 js 파일의 함수(export)

lotto step1에서 아놀드와 빅터는 이러한 생각에서 출발하여 프로젝트를 설계한 것으로 알고 있어요.

하지만, 나중에 기능이 확장될 수 있음을 가정하면 class를 사용하지 않는 설계는 한계가 있을 것 같아요.
인스턴스를 추가로 생성할 수 없으니까요.
그래서 만약 "여러 라운드의 로또 생성 내역을 히스토리로 보여줘야 할 니즈"가 생길 경우, 해결하는 데 곤란하지 않을까요?

반면, 클래스를 사용했다면, 라운드마다 모델에 새 인스턴스를 생성하고, 히스토리를 보여줄 때 여러 인스턴스의 데이터를 보여주는 방식으로 쉽게 기능 확장이 가능할 것 같아요.

----
전역변수와 함수를 이용하는 방법 외에도 조금 더 진보하여 생성자함수와 prototype을 활용하여 클래스 역할을 하는 함수를 만들어서 클래스를 대체할 수도 있겠네요.
그런데 이 방법도 클래스와 기능상의 차이점이 있다고 합니다. 자세한 내용은 다음 링크에서 확인할 수 있습니다.

[클래스는 단순한 편의 문법이 아닙니다](https://ko.javascript.info/class#ref-1343)
